### ISSUE TITLE
No prices found for `google_compute_instance` - using $0.00

### ISSUE THREAD

**Issue by cloudymax** [infracost/infracost#2437]
Hey Team, thanks so much for making infracost - it's been a huge help to my clients and myself over the years ❤️ 

## Issue

I'm having trouble getting a price for the `g2-standard` series of VMs from GCP when I run `infracost breakdown` in my terraform project https://github.com/cloudymax/gcp-tf-starter.  It seems like this is related to #2420, #2296, #2091 and a few others potentially so I'll add mine to the pile.

I see that many of the other issues had problems with reproducibility,  so I've added a github action that also generates the error reliably below. The whole project is open source and has no limits on sharing, so I'm happy to provide as much detail as possible to help move this forward 💯 

## Terraform Code

the module for the virtual machine specifically is here: https://github.com/cloudymax/modules-gcp-tf-vm. 
 
 ```hcl
resource "google_compute_instance" "compute_instance" {
  name           = random_pet.vm_name.id
  machine_type   = var.machine_type
  zone           = var.main_availability_zone
  can_ip_forward = true

  scheduling {
    preemptible                 = true
    automatic_restart           = false
    provisioning_model          = var.provisioning_model
    instance_termination_action = var.instance_termination_action
  }

  guest_accelerator {
    type  = var.guest_accelerator
    count = var.guest_accelerator_count
  }

  boot_disk {
    # kms_key_self_link = module.gcp-tf-base.google_kms_crypto_key_id
    initialize_params {
      image = var.os_image
      type  = var.disk_type
      size  = var.disk_size
    }
  }

  network_interface {
    network = "default"
    access_config {
      nat_ip = google_compute_address.static_ip.address
    }
  }

  service_account {
    # Google recommends custom service accounts that have cloud-platform scope and permissions granted via IAM Roles.
    email  = var.big_robot_email
    scopes = ["cloud-platform"]
  }

  metadata = {
    user-data = sensitive(file(var.userdata))
  }

  tags = ["web", "ssh"]
}
```

## Breakdown output:

```bash
time="2023-04-27T18:54:22Z" level=info msg="Evaluating Terraform directory at ."
time="2023-04-27T18:54:22Z" level=info msg="Starting: Downloading Terraform modules"
time="2023-04-27T18:54:23Z" level=info msg="Starting: Evaluating Terraform directory"
time="2023-04-27T18:54:23Z" level=info msg="Starting: Syncing usage data from cloud"
time="2023-04-27T18:54:23Z" level=info msg="synced 0 of 16 resources"
time="2023-04-27T18:54:23Z" level=info msg="Starting: Downloading Terraform modules"
time="2023-04-27T18:54:23Z" level=info msg="Starting: Evaluating Terraform directory"
time="2023-04-27T18:54:23Z" level=info msg="Starting: Retrieving cloud prices to calculate costs"
time="2023-04-27T18:54:23Z" level=warning msg="No prices found for module.gcp-tf-vm.google_compute_instance.compute_instance Instance usage (Linux/UNIX, preemptible, g2-standard-8), using 0.00"

Project: cloudymax/gcp-tf-starter

 Name                                                              Monthly Qty  Unit                    Monthly Cost 
                                                                                                                     
 module.gcp-tf-base.google_kms_crypto_key.key                                                                        
 ├─ Key versions                                                         25.92  months                         $1.56 
 └─ Operations                                               Monthly cost depends on usage: $0.03 per 10k operations 
                                                                                                                     
 module.gcp-tf-vm.google_compute_address.static_ip                                                                   
 └─ IP address (preemptible VM)                                            730  hours                          $1.46 
                                                                                                                     
 module.gcp-tf-vm.google_compute_instance.compute_instance                                                           
 ├─ Instance usage (Linux/UNIX, preemptible, g2-standard-8)                730  hours                          $0.00 
 └─ SSD provisioned storage (pd-ssd)                                       128  GB                            $23.94 
                                                                                                                     
 OVERALL TOTAL                                                                                                $26.95 
──────────────────────────────────
14 cloud resources were detected:
∙ 3 were estimated, 2 of which include usage-based costs, see https://infracost.io/usage-file
∙ 11 were free:
  ∙ 6 x google_compute_firewall
  ∙ 1 x google_compute_network
  ∙ 1 x google_compute_shared_vpc_host_project
  ∙ 1 x google_compute_subnetwork
  ∙ 1 x google_kms_key_ring_iam_policy
  ∙ 1 x google_service_account_key
```

## Github action to reproduce

```yaml
# The GitHub Actions docs (https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#on)
# describe other options for 'on', 'pull_request' is a good default.
name: Infracost
on: 
  pull_request:
  workflow_dispatch:
env:
  # If you use private modules you'll need this env variable to use
  # the same ssh-agent socket value across all jobs & steps.
  SSH_AUTH_SOCK: /tmp/ssh_agent.sock
jobs:
  infracost:
    name: Infracost
    runs-on: ubuntu-latest
    permissions:
      contents: read
      pull-requests: write
    env:
      TF_ROOT: .
    steps:
      - uses: actions/checkout@v3
      - name: Setup Infracost
        uses: infracost/actions/setup@v2
        # See https://github.com/infracost/actions/tree/master/setup for other inputs
        # If you can't use this action, see Docker images in https://infracost.io/cicd
        with:
          api-key: ${{ secrets.INFRACOST_API_KEY }}

      # Generate Infracost JSON file as the baseline.
      - name: Generate Infracost diff
        run: |
          infracost breakdown --path=${TF_ROOT} \
                          --show-skipped \
                          --sync-usage-file \
                          --usage-file=infracost-usage.yaml \
                          --terraform-var organization='${{ secrets.ORGANIZATION }}' \
                          --terraform-var organization_id='${{ secrets.ORGANIZATION_ID }}' \
                          --terraform-var billing_account='${{ secrets.BILLING_ACCOUNT }}'
                          
```

## Confirming that prices exist in my region/AZ
<img width="929" alt="Screenshot 2023-04-28 at 07 31 58" src="https://user-images.githubusercontent.com/84841307/235062604-0404990a-185d-4813-8642-2848619d5cab.png">

## Notes that may be helpful? IDK 🤷 

My golang skills are rubbish - _but_ - I have tried to figure out their pricing API before. I'm sharing those notes in-case my fever-dream of bash & jq  scribbles can be helpful in fixing the issue:

### How to get Prices

1. have an existing project (an organzation alone isnt enough)
2. Enable the Cloud Billing API: https://console.cloud.google.com/flows/enableapi?apiid=cloudbilling.googleapis.com
3. Create an API key: https://cloud.google.com/docs/authentication/api-keys

Create an API key using the gcloud CLI

```bash
gcloud alpha services api-keys create --display-name=SOME_NAME
```

Get price per CPU core

```bash

# JQ explanation
# 1. set the REGION arg to an array represented as a string - required to check multiple regions
# 2. filter results for items in the desired regions. uses 'index' because we need to find an item in a nested array
# 3. filter for items priced by the hour 'h'. This removes reserved instances from results
# 4. filter based on the `resourceGroups` field which is poorly named. It's closer to `family` or `machine type` from other cloud providers.
# 5. filter usage types for Preemptable only, you could also filter for OnDemand type instances.
# 6. Check if the description contains the CPU type we want. This data is oddly not in its own field anywhere.

curl https://cloudbilling.googleapis.com/v1/services/6F81-5844-456A/skus?key=$(bw get notes "GCP API key") > skus_compute_engine.json

export FAMILY="N1Standard" # or `CPU`
export REGION='"europe-west1", "europe-west4", "europe-central2"'
export CORES="16"
export CPU_TIER="N1" # or A2 

DATA=$(cat skus_compute_engine.json \
  | jq -r --arg REGION "$REGION" '.skus[] 
  | select((.serviceRegions | index( '"$REGION"' )) 
  and select(.pricingInfo[0].pricingExpression.usageUnit=="h") 
  and .category.resourceGroup==env.FAMILY 
  and .category.usageType=="Preemptible" 
  and select(.description | contains( env.CPU_TIER )))')

export NANOS=$(echo $DATA |jq '.pricingInfo[0].pricingExpression.tieredRates[0].unitPrice.nanos')
CONVERTED_RATE=$(bc <<< "scale=5; $NANOS/1000000000")
CPU_PRICE=$( bc <<< "scale=5; $CONVERTED_RATE * $CORES" )
echo "CPU Price: $CPU_PRICE"
```

Get RAM price:

```bash
export FAMILY="N1Standard" # or `RAM`
export REGION='"europe-west1", "europe-west4"'
export RAM_AMOUNT="64"
export CPU_TIER="N1"

DATA=$(cat skus_compute_engine.json \
  | jq -r --arg REGION "$REGION" '.skus[] 
  | select((.serviceRegions | index( '"$REGION"' )) 
  and select(.pricingInfo[0].pricingExpression.usageUnit=="GiBy.h") 
  and .category.resourceGroup==env.FAMILY 
  and .category.usageType=="Preemptible" 
  and select(.description | contains( env.CPU_TIER )))')

export NANOS=$(echo $DATA |jq '.pricingInfo[0].pricingExpression.tieredRates[0].unitPrice.nanos')
CONVERTED_RATE=$(bc <<< "scale=5; $NANOS/1000000000")
RAM_PRICE=$( bc <<< "scale=5; $CONVERTED_RATE * $RAM_AMOUNT" )
echo "RAM Price: $RAM_PRICE"
```

Get price per GPU

```bash
# JQ explanation
# 1. set the REGION arg to an array represented as a string - required to check multiple regions
# 2. filter results for items in the desired regions. uses 'index' because we need to find an item in a nested array
# 3. filter for items priced by the hour 'h'. This removes reserved instances from results
# 4. filter based on the `resourceGroupz filed which is poorly named. It's closer to 'family' or 'machine type' from other cloud providers.
# 5. filter usage types for Preemptable only, you could also filter for OnDemand type instances.
# 6. Check if the description contains the GPU type we want. This data is oddly not in its own filed anywhere.

GPU Types: T4, P4, A100, P100

export FAMILY="GPU"
export GPU_TYPE="T4"
export GPUS=1
export REGION='"europe-west1", "europe-west4"'

DATA=$(cat skus_compute_engine.json \
  | jq -r --arg REGION "$REGION" '.skus[] 
  | select((.serviceRegions | index( '"$REGION"' )) 
  and select(.pricingInfo[0].pricingExpression.usageUnit=="h") 
  and .category.resourceGroup==env.FAMILY 
  and .category.usageType=="Preemptible" 
  and select(.description | contains( env.GPU_TYPE )))')


export NANOS=$(echo $DATA \
  |jq '.pricingInfo[0].pricingExpression.tieredRates[0].unitPrice.nanos')

CONVERTED_RATE=$(bc <<< "scale=5; $NANOS/1000000000")

GPU_PRICE=$( bc <<< "scale=5; $CONVERTED_RATE * $GPUS" )

echo "GPU Price: $GPU_PRICE"
```

Combine prices

```bash
COMBINED_PRICE=$(bc <<< "scale=5; $CPU_PRICE + $RAM_PRICE + $GPU_PRICE")
echo "Combined Price: $COMBINED_PRICE"
```

**Comment by aliscott** [comment]
Thanks for the detailed issue @cloudymax. This is the level of detail that helps us quickly reproduce and fix these issues! 🙂 

I've been able to reproduce with the following Terraform code:
```
provider "google" {
  credentials = "{\"type\":\"service_account\"}"
  region      = "us-central1"
}

resource "google_compute_instance" "standard" {
  name         = "standard"
  machine_type = "g2-standard-12"
  zone         = "us-central1-a"

  boot_disk {
    initialize_params {
      image = "centos-cloud/centos-7"
    }
  }

  network_interface {
    network = "default"
  }
}
```

It looks like empty prices are coming through to our pricing DB for all `g2-standard` instances, so the issue might be in This

```
> SELECT prices FROM products where "vendorName"='gcp' AND "service"='Compute Engine' AND "productFamily"='Compute Instance'  AND "region"='us-central1' AND "attributes"->>'machineType' LIKE 'g2-standard%';
 prices
--------
 {}
 {}
 {}
 {}
 {}
 {}
 {}
 {}
```

This doesn't appear to be the same issue as #2420, but might be related. We will look at getting a fix in for the next release.

**Comment by cloudymax** [comment]
Thanks for your reply @aliscott!

I'll keep my eyes open for the next release then - cheers!

**Comment by aliscott** [comment]
Thanks for report this @cloudymax 🙏 . This has now been fixed and released in [v0.3.15 of the Cloud Pricing API](https://github.com/infracost/cloud-pricing-api/releases/tag/v0.3.15).

**Comment by cloudymax** [comment]
Awesome work! Thanks again 🥳 
<img width="1242" alt="Screenshot 2023-05-03 at 20 32 16" src="https://user-images.githubusercontent.com/84841307/236011339-c01edf8a-684b-4980-8c5d-20d6c210b627.png">
